src/network/core/tcp_connect.cpp

```diff
@@ -68,32 +298,22 @@ void TCPConnecter::Connect()
 {
 	for (auto iter = _tcp_connecters.begin(); iter < _tcp_connecters.end(); /* nothing */) {
 		TCPConnecter *cur = *iter;
-		const bool connected = cur->connected.load();
-		const bool aborted = cur->aborted.load();
-		if ((connected || aborted) && cur->killed) {
-			iter = _tcp_connecters.erase(iter);
-			if (cur->sock != INVALID_SOCKET) closesocket(cur->sock);
-			delete cur;
-			continue;
-		}
-		if (connected) {
-			iter = _tcp_connecters.erase(iter);
-			cur->OnConnect(cur->sock);
-			delete cur;
-			continue;
-		}
-		if (aborted) {
+
+		if (cur->CheckActivity()) {
 			iter = _tcp_connecters.erase(iter);
-			cur->OnFailure();
 			delete cur;
-			continue;
+		} else {
+			iter++;
 		}
-		iter++;
 	}
 }
 
 /** Kill all connection attempts. */
 /* static */ void TCPConnecter::KillAll()
 {
-	for (TCPConnecter *conn : _tcp_connecters) conn->killed = true;
+	for (auto iter = _tcp_connecters.begin(); iter < _tcp_connecters.end(); /* nothing */) {
```

Comment:
Won't a `_tcp_connecters.clear()` not do the trick? Or does it then need to become a vector<unique_ptr<TCPConnecter>>?

Reply:
Exactly what you say; something first needs to delete the object, so just running `clear()` means it would leak.

`std::unique_ptr` isn't trivial, as the entries are added to the list by their ctor. So I left it alone for now.